### ISSUE TITLE
rustc overflows its stack when type checking ~~a simple program~~ polymorphmic recursion

### ISSUE THREAD

**Issue by jhjourdan** [rust-lang/rust#38591]
The following program triggers a stack overflow in rustc:

```rust
struct S<T> {
    t : T,
    s : Box<S<fn(u : T)>>
}

fn f(x : S<u32>) { }

fn main () { }
```

My version of rustc:

```
$ rustc --version
rustc 1.15.0-nightly (71c06a56a 2016-12-18)
```

The problem also appears when replacing `fn(u : T)` with `fn() -> T`.

**Comment by achanda** [comment]
Interestingly, I don't see a stack trace with `RUST_BACKTRACE`
```
:~/src$ RUST_BACKTRACE=1 rustc test.rs

thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
```
Same version of `rustc`

**Comment by AraHaan** [comment]
Well oddly than not it does look to see if there is about half a GB left on disk and if there is not updating would not work. Which is bad because there actually is enough space if I have an older version of it.

**Comment by theduke** [comment]
The struct is obviously infinitely recursive.

The compiler should catch that of course and emit something like "illegal recursive struct" or whatever.

**Comment by AraHaan** [comment]
This also happens that sometimes it actually uses over than 1 GB of RAM as well for some insanely small code for some unknown reason.

**Comment by jhjourdan** [comment]
> The struct is obviously infinitely recursive.

I don't know what you mean by this, but I do no think this is the problem here. The following program also reveals the problem:

```
struct S<T> {
    t : T,
    s : Box<S<Option<fn(u : T)>>>
}
```

But the following is accepted:
```
struct S<T> {
    t : T,
    s : Box<S<T>>
}

fn f(x : S<u32>) { }

fn main () { }
```
Recursive structs are a well-known and important feature of Rust.

**Comment by sfackler** [comment]
@jhjourdan `S<()>` contains an `S<fn(u: ())>`, which contains an `S<fn(u: fn(u: ()))>`, which contains an `S<fn(u: fn(u: fn(u: ())))>`, which contains...

**Comment by jhjourdan** [comment]
@sfackler : it does not /contain/ it. It /refers/ to it.

Again, recursive structs are a well-known, important feature of Rust. For example, you cannot define linked lists without them.

**Comment by sfackler** [comment]
The issue is not the storage of the struct. The issue is that you are asking the compiler to generate code for an infinite number of instantiations of `S`.

```rust
struct List<T>(Option<Box<List<T>>>)
```

Is a very different thing than

```rust
struct Foo<T>(Option<Box<List<fn(T)>>>);
```

**Comment by jhjourdan** [comment]
Well, but I am not asking the compiler to generate any code here... Or at least, the code that is generated is not used anywhere.

**Comment by sfackler** [comment]
I'm confused then - why is it valuable for it to be possible to define but never use this type?

**Comment by jhjourdan** [comment]
Btw, now I understand better why my code provokes a stack overflow.

But still, there is no reason this code is generated if the code is never used, and even then, this is not an excuse for rustc to do a stack overflow.

**Comment by brson** [comment]
Yes there should be some reasonable recursion limit here.

**Comment by insaneinside** [comment]
More importantly, rustc should issue a diagnostic instead of overflowing its stack. 😮

**Comment by istankovic** [comment]
This is no longer reproducible with rustc 1.26.0-nightly (2789b067d 2018-03-06).

**Comment by pnkfelix** [comment]
The phenomenon that @sfackler is describing [above][] is technically known as "polymorphic recursion."

[above]: https://github.com/rust-lang/rust/issues/38591#issuecomment-269128142

cc #4287

But, as noted by @istankovic above, this particular instance of the diagnostic issue with polymorphic recursion (that I believe persists to this day) no longer seems to replicate.

Marking as E-needstest.

**Comment by nu8** [comment]
https://github.com/rust-lang/rust/issues/52852
https://github.com/rust-lang/rust/issues/4287
